gapic/src/main/com/google/gapid/views/DeviceValidationView.java

```diff
@@ -103,46 +136,62 @@ public void ValidateDevice(Device.Instance dev) {
-      protected void onUiThreadSuccess(DeviceValidationResult result) {
-        setValidationStatus(result);
-      }
 
-      @Override
-      protected void onUiThreadError(DeviceValidationResult result) {
-        LOG.log(WARNING, "UI thread error while validating device support");
-        setValidationStatus(result);
-      }
+    // Assign appropriate callback for retry button.
+    for (Listener listener : retryButton.getListeners(SWT.Selection)) {
+      retryButton.removeListener(SWT.Selection, listener);
+    }
+    retryButton.addListener(SWT.Selection, e -> {
+      ValidateDevice(dev);
     });
+
+    models.devices.validateDevice(dev, this::setValidationStatus);
   }
 
   private void setValidationStatus(DeviceValidationResult result) {
-    if (result.skipped) {
-      statusLoader.updateStatus(true);
-      statusLoader.stopLoading();
-      statusText.setText("Device support validation skipped.");
-      validationPassed = true;
-    } else {
-      statusLoader.updateStatus(result.passed);
-      statusLoader.stopLoading();
-      statusText.setText("Device support validation " 
-          + (result.passed ? "passed." : "failed. " + Messages.VALIDATION_FAILED_LANDING_PAGE));
-      validationPassed = result.passed;
-    }
+    boolean passedOrSkipped = result.passed || result.skipped;
+    statusLoader.stopLoading();
+    statusLoader.updateStatus(passedOrSkipped);
+    validationPassed = passedOrSkipped;
+    statusText.setText("Device support validation " + result.toString() + ".");
     notifyListeners(SWT.Modify, new Event());
+
+    if (passedOrSkipped) {
+      return;
+    }
+
+    // Extra details (i.e. error message & help text)
+    retryButton.setVisible(true);
+
+    errorMessageGroup = withLayoutData(createGroup(this, ""),
+      withSpans(new GridData(SWT.FILL, SWT.TOP, true, false), 3, 1));
+    Text errText = withLayoutData(
+      createTextbox(errorMessageGroup, SWT.READ_ONLY | SWT.WRAP , result.validationFailureMsg),
+      new GridData(SWT.LEFT, SWT.TOP, false, false));
+
+    helpComposite = createComposite(this, new GridLayout(1, false));
+    helpComposite.setLayoutData(withSpans(new GridData(), 3, 1));
+    Link helpLink = withLayoutData(createLink(helpComposite, Messages.VALIDATION_FAILED_LANDING_PAGE, e -> {
+      Program.launch(URLs.DEVICE_COMPATIBILITY_URL);
+    }), new GridData(SWT.LEFT, SWT.TOP, false, false));
+    if (result.tracePath.length() > 0) {
+      Link traceLink = withLayoutData(createLink(helpComposite, "View <a>trace file</a>", e -> {
+        try {
+          OS.openFileInSystemExplorer(new File(result.tracePath));
+        } catch (IOException exception) {
+          LOG.log(SEVERE, "Failed to open log directory in system explorer", exception);
+        }
+      }), new GridData(SWT.LEFT, SWT.TOP, false, false));
+    }
+
+    // Resize the rest of the window if needed.
+    Point curr = getShell().getSize();
+    Point want = getShell().computeSize(SWT.DEFAULT, SWT.DEFAULT);
```

Comment:
Since you will not resize the X dimension, pass it to `computeSize`: `getShell().computeSize(curr.x, SWT.DEFAULT);`